### PR TITLE
Impute age categories

### PR DESCRIPTION
**Pull request by IsaackMwenda**
TODO: plug-in sync stats in a future pr

### CODE REVIEW

File: src/engagement_db_to_analysis/data_wrangling_functions.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    # Get the configurations for age and age_category analysis datasets
+    updated_messages_traced_data = []
+    age_cc = None
+    age_category_cc = None
+    age_engagement_db_datasets = None
+    for analysis_dataset_config in analysis_dataset_configs:
+        for coding_config in analysis_dataset_config.coding_configs:
+            if coding_config.analysis_dataset == "age":
+                age_cc = coding_config
+                age_engagement_db_datasets = analysis_dataset_config.engagement_db_datasets
```
Comment (DanielMwendwa) [inline src/engagement_db_to_analysis/data_wrangling_functions.py:33]: I suggest renaming `age_engagement_db_datasets` to `analysis_age_dataset_config` and then update L33 to `analysis_age_dataset_config = analysis_dataset_config` so that the code block does what you've commented in L24. We can then get age engagement db datasets by `analysis_age_dataset_config.engagement_db_datasets`

File: src/engagement_db_to_analysis/data_wrangling_functions.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            elif coding_config.analysis_dataset == "age_category":
+                age_category_cc = coding_config
+
+    #Check and impute age_category in age messages only
+    for message in messages_traced_data:
```
Comment (DanielMwendwa) [inline src/engagement_db_to_analysis/data_wrangling_functions.py:38]: move `updated_messages_traced_data = []` on top of this for loop.

File: src/engagement_db_to_analysis/data_wrangling_functions.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        if message["dataset"] not in age_engagement_db_datasets:
+            updated_messages_traced_data.append(message)
+            continue
```
Comment (DanielMwendwa) [inline src/engagement_db_to_analysis/data_wrangling_functions.py:41]: You can check for `message["dataset"] in age_engagement_db_datasets` and then update messages traced data at the end of the for loop.

File: src/engagement_db_to_analysis/data_wrangling_functions.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+        # Append this age_category_label to the list of labels for this message, and write-back to TracedData.
+        message_labels = message["labels"].copy()
+        message_labels.append(age_category_label.to_dict())
+        message.append_data(
+            {"labels": message_labels},
+            Metadata(user, Metadata.get_call_location(), TimeUtils.utc_now_as_iso_string())
+        )
+
+        updated_messages_traced_data.append(message)
```
Comment (DanielMwendwa) [inline src/engagement_db_to_analysis/data_wrangling_functions.py:73]: Should we here also append trace data to message with message history?

**Review by DanielMwendwa — APPROVED** [review]

File: configurations/test_pipeline_configuration.py [hunk 5]
```diff
@@ -142,7 +142,14 @@
                     ),
                     CodingConfiguration(
                         code_scheme=load_code_scheme("age_category"),
-                        analysis_dataset="age_category"
+                        analysis_dataset="age_category",
+                        age_categories = {
```
Comment (as2388) [inline configurations/test_pipeline_configuration.py:146]:
```suggestion
                        age_categories={
```

File: configurations/test_pipeline_configuration.py [hunk 6]
```diff
@@ -142,7 +142,14 @@
...
+                            (10, 14): "10 to 14",
+                            (15, 17): "15 to 17",
+                            (18, 35): "18 to 35",
+                            (36, 54): "36 to 54",
+                            (55, 99): "55 to 99"
+                        }
```
Comment (as2388) [inline configurations/test_pipeline_configuration.py:152]: I think this is ok and happy to keep like this so we can keep making progress.

The slight hesitations I have are that
i. this means we need to specify the categories every time, rather than relying on an avf default.
ii. the functional programmer in me instinctively leans to something like code_imputations=[ImputeAgeCategories(categories=...), ...] as it makes adding additional imputations easier, but that might well be overkill here

File: src/engagement_db_to_analysis/data_wrangling_functions.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+    :param user: Identifier of user running the pipeline.
+    :type user: str
+    :param messages_traced_data: Messages TracedData objects to impute age_category.
+    :type messages_traced_data: list of TracedData
+    :param analysis_dataset_config: Analysis dataset configuration in pipeline configuration module.
+    :type analysis_dataset_config: pipeline_config.analysis_configs.dataset_configurations
+    """
+
+    log.info("Imputing age categories for age dataset messages...")
```
Comment (as2388) [inline src/engagement_db_to_analysis/data_wrangling_functions.py:23]: I think we should say what the dataset actually is, as it might not exactly be called 'age'

File: src/engagement_db_to_analysis/data_wrangling_functions.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
+from core_data_modules.util import TimeUtils
```
Comment (as2388) [inline src/engagement_db_to_analysis/data_wrangling_functions.py:1]: Can we call this file code_imputation_functions instead? I think all of analysis can arguably be described as a data wrangling operation

File: src/engagement_db_to_analysis/data_wrangling_functions.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+    log.info("Imputing age categories for age dataset messages...")
+    # Get the configurations for age and age_category analysis datasets
+    updated_messages_traced_data = []
+    age_cc = None
+    age_category_cc = None
+    age_engagement_db_datasets = None
+    for analysis_dataset_config in analysis_dataset_configs:
+        for coding_config in analysis_dataset_config.coding_configs:
+            if coding_config.analysis_dataset == "age":
```
Comment (as2388) [inline src/engagement_db_to_analysis/data_wrangling_functions.py:31]: I'm not sure about this because it means we always need to call these datasets age and age_category, which might not be true and it's not clear when writing a configuration file that you need to do this.

I think I see the problem though, which is that the age_categorisation needs to know where to look to find the age dataset - maybe change the configuration to something like
age_category=AgeCategoryConfiguration(age_dataset="age", [categories=...])

File: src/engagement_db_to_analysis/data_wrangling_functions.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+            age_category_code = age_category_cc.code_scheme.get_code_with_control_code(
+                age_code.control_code)
+
+        age_category_label = CleaningUtils.make_label_from_cleaner_code(
+            age_category_cc.code_scheme, age_category_code, Metadata.get_call_location()
+        )
+
+        # Append this age_category_label to the list of labels for this message, and write-back to TracedData.
+        message_labels = message["labels"].copy()
+        message_labels.append(age_category_label.to_dict())
```
Comment (as2388) [inline src/engagement_db_to_analysis/data_wrangling_functions.py:67]: Labels are sorted newest first, so I think you want to append at the front of this list

**Review by as2388 — CHANGES_REQUESTED** [review]

File: configurations/test_pipeline_configuration.py (hunk 6 above)
Comment (IsaackMwenda) [reply configurations/test_pipeline_configuration.py:152]: on i)  We've seen this vary between projects in Cesc depending on the target audience. Moving this here standardizes _impute_age_category() so we update the categories & scheme between projects.
ii) My view is storing imputation configs in their resp coding_plan makes it easier to trace computations that happen in each analysis dataset.

File: src/engagement_db_to_analysis/data_wrangling_functions.py (hunk 1 above)
Comment (IsaackMwenda) [reply src/engagement_db_to_analysis/data_wrangling_functions.py:33]: So we only need the coding_config and age_engagement_db_datasets for the rest of the script and not really analysis_age_dataset_config. So have updated the comment in L24 to be more clear :).

File: src/engagement_db_to_analysis/data_wrangling_functions.py (hunk 2 above)
Comment (IsaackMwenda) [reply src/engagement_db_to_analysis/data_wrangling_functions.py:38]: Updated thanks.

File: src/engagement_db_to_analysis/data_wrangling_functions.py (hunk 3 above)
Comment (IsaackMwenda) [reply src/engagement_db_to_analysis/data_wrangling_functions.py:41]: updated thanks.

File: src/engagement_db_to_analysis/data_wrangling_functions.py (hunk 4 above)
Comment (IsaackMwenda) [reply src/engagement_db_to_analysis/data_wrangling_functions.py:73]: Not really because;
1. updated_messages_traced_data is just a normal list that stores the traced data objects.
2. We've attached history for every message we have mutated in L67.

File: src/engagement_db_to_analysis/data_wrangling_functions.py (hunk 7 above)
Comment (IsaackMwenda) [reply src/engagement_db_to_analysis/data_wrangling_functions.py:23]: updated thanks

File: src/engagement_db_to_analysis/data_wrangling_functions.py (hunk 10 above)
Comment (IsaackMwenda) [reply src/engagement_db_to_analysis/data_wrangling_functions.py:67]: updated thanks.

File: src/engagement_db_to_analysis/data_wrangling_functions.py (hunk 9 above)
Comment (IsaackMwenda) [reply src/engagement_db_to_analysis/data_wrangling_functions.py:30]: It's not just that we also need to loop through age messages only, and to determine this we have to check for an age config property in each message. The alternative here is to define constants e.g AGE_CATEGORY = "age_category" in configuration.py that we can then use in the scripts in place of strings.

File: configurations/test_pipeline_configuration.py [hunk 11]
```diff
@@ -142,9 +142,19 @@
...
+                        analysis_dataset="age_category",
+                        age_category_config=AgeCategoryConfiguration(age_analysis_dataset="age", #age_cc.analysis_dataset
+                                                                     categories={
+                                                                         (10, 14): "10 to 14",
+                                                                         (15, 17): "15 to 17",
+                                                                         (18, 35): "18 to 35",
+                                                                         (36, 54): "36 to 54",
+                                                                         (55, 99): "55 to 99"
+                                                                     }
+                                                                     )
```
Comment (as2388) [inline configurations/test_pipeline_configuration.py:154]:
```suggestion
                        age_category_config=AgeCategoryConfiguration(
                            age_analysis_dataset="age",
                            categories={
                                (10, 14): "10 to 14",
                                (15, 17): "15 to 17",
                                (18, 35): "18 to 35",
                                (36, 54): "36 to 54",
                                (55, 99): "55 to 99"
                             }
                      )
```

File: src/engagement_db_to_analysis/code_imputation_functions.py [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+    :param analysis_dataset_configs: Analysis dataset configuration in pipeline configuration module.
+    :type analysis_dataset_configs: pipeline_config.analysis_configs.dataset_configurations
+    """
+
+    # Get the coding configurations for age and age_category analysis datasets
+    age_category_cc = None
+    for analysis_dataset_config in analysis_dataset_configs:
+        for coding_config in analysis_dataset_config.coding_configs:
+            if coding_config.age_category_config is not None:
+                age_category_cc = coding_config
```
Comment (as2388) [inline src/engagement_db_to_analysis/code_imputation_functions.py:28]: This will fail if we don't provide an age category dataset, and have unexpected results if we provide it twice. I think you need to return early (with a log) if there is no dataset where this was requested, and fail if you see it twice

File: src/engagement_db_to_analysis/code_imputation_functions.py [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+
+    age_cc = None
```
Comment (as2388) [inline src/engagement_db_to_analysis/code_imputation_functions.py:30]: Maybe this should be age_coding_config.

age_cc was fine when it was immediately preceded by for age_cc in coding_configurations:, but it's not obvious what the cc stands for without that for loop

File: src/engagement_db_to_analysis/code_imputation_functions.py [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+    age_engagement_db_datasets = None
+    for analysis_dataset_config in analysis_dataset_configs:
+        for coding_config in analysis_dataset_config.coding_configs:
+            if coding_config.analysis_dataset == age_category_cc.age_category_config.age_analysis_dataset:
```
Comment (as2388) [inline src/engagement_db_to_analysis/code_imputation_functions.py:34]: again, what happens if there were multiple matches. I think we need to check here (and/or pre-validate when the config gets loaded)

File: src/engagement_db_to_analysis/code_imputation_functions.py [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+                age_cc = coding_config
+                age_engagement_db_datasets = analysis_dataset_config.engagement_db_datasets
+
+    # Check and impute age_category in age messages only
+    log.info(f"Imputing {age_category_cc.analysis_dataset} for {age_cc.analysis_dataset} messages...")
```
Comment (as2388) [inline src/engagement_db_to_analysis/code_imputation_functions.py:39]:
```suggestion
    log.info(f"Imputing {age_category_cc.analysis_dataset} from {age_cc.analysis_dataset} messages...")
```

File: src/engagement_db_to_analysis/code_imputation_functions.py [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+def _impute_age_category(user, messages_traced_data, analysis_dataset_configs):
+    """
+    Imputes age category for age dataset messages.
+
+    :param user: Identifier of user running the pipeline.
+    :type user: str
+    :param messages_traced_data: Messages TracedData objects to impute age_category.
+    :type messages_traced_data: list of TracedData
+    :param analysis_dataset_configs: Analysis dataset configuration in pipeline configuration module.
+    :type analysis_dataset_configs: pipeline_config.analysis_configs.dataset_configurations
```
Comment (as2388) [inline src/engagement_db_to_analysis/code_imputation_functions.py:20]: Please doc the return (as with TracedData, we have to be super clear about when we're appending in place and when we're returning a modified copy)

File: src/engagement_db_to_analysis/code_imputation_functions.py [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+            if coding_config.analysis_dataset == age_category_cc.age_category_config.age_analysis_dataset:
+                age_cc = coding_config
+                age_engagement_db_datasets = analysis_dataset_config.engagement_db_datasets
+
+    # Check and impute age_category in age messages only
+    log.info(f"Imputing {age_category_cc.analysis_dataset} for {age_cc.analysis_dataset} messages...")
+    updated_messages_traced_data = []
+    for message in messages_traced_data:
+        if message["dataset"] in age_engagement_db_datasets:
+            age_label = message["labels"][0]
```
Comment (as2388) [inline src/engagement_db_to_analysis/code_imputation_functions.py:43]: What if we. have multiple labels assigned to this message? I think you need to get the latest label(s) with this code scheme - there is a function in column_view_conversion that handles this

File: src/engagement_db_to_analysis/code_imputation_functions.py [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+            message_labels = message["labels"].copy()
+            message_labels.insert(0, age_category_label.to_dict())
+            message.append_data(
+                {"labels": message_labels},
+                Metadata(user, Metadata.get_call_location(), TimeUtils.utc_now_as_iso_string())
+            )
+
+        updated_messages_traced_data.append(message)
+
+        return updated_messages_traced_data
```
Comment (as2388) [inline src/engagement_db_to_analysis/code_imputation_functions.py:75]: This means you're updating the message in place but returning a new list which is confusing. I think you either need to copy everything, which will be noticeably slow, or append in place, which I think is the better option here.

File: src/engagement_db_to_analysis/engagement_db_to_analysis.py [hunk 19]
```diff
@@ -159,6 +160,7 @@ def generate_analysis_files(user, pipeline_config, engagement_db, engagement_db_
 
     messages_traced_data = filter_messages(user, messages_traced_data, pipeline_config)
 
-    participants_traced_data_map = _fold_messages_by_uid(user, messages_traced_data)
+    messages_traced_data = run_code_computation_functions(user, messages_traced_data,
+                                                        pipeline_config.analysis_configs.dataset_configurations)
```
Comment (as2388) [inline src/engagement_db_to_analysis/engagement_db_to_analysis.py:164]:
```suggestion
    messages_traced_data = run_code_computation_functions(
        user, messages_traced_data, pipeline_config.analysis_configs.dataset_configurations
    )
```
In general I think this is the better way to indent, because otherwise you end up with lots of very short lines at the far right of the screen

File: src/pipeline_configuration_spec.py [hunk 20]
```diff
@@ -45,11 +45,11 @@ class PipelineConfiguration:
...
     uuid_table: UUIDTableClientConfiguration
-    analysis_configs: AnalysisConfiguration
     project_start_date: datetime = None
     project_end_date: datetime = None
     test_participant_uuids: [] = None
     description: str = None
     rapid_pro_sources: [RapidProSource] = None
     coda_sync: CodaConfiguration = None
     rapid_pro_target: RapidProTarget = None
+    analysis_configs: AnalysisConfiguration = None
```
Comment (as2388) [inline src/pipeline_configuration_spec.py:55]: (drive-by comment, perhaps for another PR - This only holds one AnalysisConfiguration, so should be analysis_config not analysis_configs)

File: src/engagement_db_to_analysis/configuration.py [hunk 21]
```diff
@@ -8,10 +9,16 @@ class DatasetTypes:
     DEMOGRAPHIC = "demographic"
     RESEARCH_QUESTION_ANSWER = "research_question_answer"
 
+@dataclass
+class AgeCategoryConfiguration:
+    age_analysis_dataset: str
+    categories: dict
+
```
Comment (as2388) [inline src/engagement_db_to_analysis/configuration.py:16]:
```suggestion


```
my editor is very noisy about PEP-8 violations without double-blank lines in modules

File: src/engagement_db_to_analysis/code_imputation_functions.py [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+
+def run_code_computation_functions(user, messages_traced_data, analysis_dataset_config):
```
Comment (as2388) [inline src/engagement_db_to_analysis/code_imputation_functions.py:77]:
```suggestion
def impute_codes_by_message(user, messages_traced_data, analysis_dataset_config):
```
or similar - basically, we'll need to distinguish imputations that run message by message and imputations that run by participant

**Review by as2388 — APPROVED** [review]

File: src/engagement_db_to_analysis/code_imputation_functions.py (hunk 12 above)
Comment (IsaackMwenda) [reply src/engagement_db_to_analysis/code_imputation_functions.py]: Added an assert if age_category_cc is more than one and log for coding configs that don't have it.

File: src/engagement_db_to_analysis/code_imputation_functions.py (hunk 13 above)
Comment (IsaackMwenda) [reply src/engagement_db_to_analysis/code_imputation_functions.py]: updated to age_coding_config

File: src/engagement_db_to_analysis/engagement_db_to_analysis.py (hunk 19 above)
Comment (IsaackMwenda) [reply src/engagement_db_to_analysis/engagement_db_to_analysis.py]: updated.

File: src/engagement_db_to_analysis/code_imputation_functions.py (hunk 22 above)
Comment (IsaackMwenda) [reply src/engagement_db_to_analysis/code_imputation_functions.py]: Updated thanks.

File: src/engagement_db_to_analysis/configuration.py (hunk 21 above)
Comment (IsaackMwenda) [reply src/engagement_db_to_analysis/configuration.py:17]: Updated thanks.

File: src/pipeline_configuration_spec.py (hunk 20 above)
Comment (IsaackMwenda) [reply src/pipeline_configuration_spec.py:55]: Thanks will update in a future PR.
